Add explicit nullability support

tfcporciuncula opened the pull request:
Inspired by https://github.com/xmartlabs/flow-preferences/pull/1/files to fix https://github.com/tfcporciuncula/flow-preferences/issues/6

- Introduces `NullableStringPreference` so users can explicitly opt in for nullability support when they're working with strings -- `StringPreference` remains strict against nullables.
- Introduces `NullableObjectPreference` and changes `ObjectPreference` to not accept nullable values to keep things consistent.
- Renames the current `NullableStringSetPreference` to `StringSetOfNullablesPreference`, and introduces a new `NullableStringSetPreference` and `NullableStringSetOfNullablesPreference` for actual `null` set support.

There's some code repetition here, but things are still pretty simple so I'm fine with it. I like how the library is now extremely explicit about nulls and I'll probably add a new section about this on the readme if this gets merged.

Commit:
Add typealias to avoid some nullability pollution

Commit:
Introduce NullableStringPreference

Commit:
Restrict ObjectPreference null support

Commit:
Introduce NullableObjectPreference

Commit:
Rename NullableStringSetPreference

To StringSetOfNullablesPreference, so we can introduce NullableStringSetPreference as a version that accepts an actual null set. We’ll also introduce NullableStringSetOfNullablesPreference…. oh boy.

Commit:
Add the other two variants of string set pref

Also fixing a typo on the existing string set tests.

Commit:
Fix static analysis

mirland commented on flow-preferences/src/main/java/com/tfcporciuncula/flow/FlowSharedPreferences.kt:
```diff
@@ -40,17 +42,40 @@ class FlowSharedPreferences @JvmOverloads constructor(
+  ): Preference<T> =
     ObjectPreference(keyFlow, sharedPreferences, key, serializer, defaultValue, coroutineContext)
 
+  fun <T> getNullableObject(
```
```suggestion
  @JvmOverloads
  fun <T> getNullableObject(
```

mirland commented on flow-preferences/src/main/java/com/tfcporciuncula/flow/FlowSharedPreferences.kt:
```diff
@@ -40,17 +42,40 @@ class FlowSharedPreferences @JvmOverloads constructor(
+  fun getNullableStringSetOfNullables(key: String, defaultValue: Set<String?>? = null): Preference<Set<String?>?> =
+    NullableStringSetOfNullablesPreference(keyFlow, sharedPreferences, key, defaultValue, coroutineContext)
+
+  fun <T : Any> getObject(
```
```suggestion
  @JvmOverloads
  fun <T : Any> getObject(
```

mirland reviewed (approved):
Great work!

Just a few comments:
- Readme documentation is missing (I know that you added it in the or description, but maybe you can include it in this pr)
- I'd consider defining `BasePreference` (and the subclasses) as `internal` classes.
- I'd document the `Preference` class, (and maybe `FlowSharedPreferences` too)
- I'd consider defining `EnumPreference` as an `ObjectPreference` with an `EnumSerializer`
- I'd consider defining `FlowSharedPreferences.keyFlow` as `internal`

Commit:
Add missing @JvmOverloads

Co-authored-by: Matías Irland <matiasirland@gmail.com>

Commit:
Add missing @JvmOverloads [2]

Co-authored-by: Matías Irland <matiasirland@gmail.com>

tfcporciuncula commented:
- 💯 agree with the `internal`, but to keep the scope smaller here I'll have it as a separate commit after we merge this.
-  I'll add the readme changes here, that makes sense.
- About documentation, yeah, it's probably time to add some. But I'll also work on it separately to avoid increasing the scope here. Depending on how important this is to you, feel free to open a separate issue to address it.
- I like the enum idea, it'd definitely make things simpler! Will also leave it as it is for now just to avoid increasing the scope here (again).

Thanks a lot for the review and for the great suggestions 🙌

Commit:
Add new readme section

tfcporciuncula commented:
@matir91 

- I created an issue to add documentation https://github.com/tfcporciuncula/flow-preferences/issues/8.
- Made things `internal` and updated the enum implementation https://github.com/tfcporciuncula/flow-preferences/commit/2d0a49ba51d540a605665bd33185ba9fa5c43da5
- And added support for nullable enums that we missed https://github.com/tfcporciuncula/flow-preferences/commit/d9a2fd5b1c03d746483303b5f85f4352c1bc2987

I think we should be good for a new release now.
